src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py

```diff
@@ -0,0 +1,194 @@
+        """
+        Return list of requests specs for provided list of request names
+        :param requests: list of workflow requests names
+        :return specs: dictionary of specs obtained from upstream ReqMgr2 server
+        """
+        urls = []
+        for name in requests:
+            url = "{}/data/request?name={}".format(self.reqmgrUrl, name)
+            urls.append(url)
+        response = getdata(urls, ckey(), cert())
+        specs = {}
+        for resp in response:
+            data = json.loads(resp['data'])
+            for rdict in data['result']:
+                specs.update(rdict)
+        return specs
+
+    @timeFunction
+    def algorithm(self, parameters=None):
+        """
+        Executed in every polling cycle. The actual logic of the component is:
+        - obtain list of current active workflows from the agent
+        - requests their specs from upstream ReqMgr2 server
+        - update site lists of all workflows
+        - push new specs to the agent local WorkQueue and update pickle spec file
+
+        :return: none
+        """
+        # get list of active workflows from the agent, the returned dict
+        # is composed by workflow names and associated pickle file (data comes from WMBS)
+        wmaSpecs = self.getActiveWorkflows()
+        wflows = wmaSpecs.keys()
+
+        # obtain workflow specs from upstream ReqMgr2 server, the returned dict
+        # is composed by workflow names and JSON dict representing the workflow
+        specs = self.getRequestSpecs(wflows)
+
+        # iterate over workflow items and update local WorkQueue and pickle files if
+        # either site white or black lists are different
+        for wflow, rdict in specs.items():
+            siteWhiteList = rdict['SiteWhitelist']
+            siteBlackList = rdict['SiteBlacklist']
+
+            # get the name of pkl file from wma spec
+            pklFileName = wmaSpecs[wflow]
+
+            # create wrapper helper and load pickle file
+            wHelper = WMWorkloadHelper()
+            wHelper.load(pklFileName)
+
+            # extract from pickle spec both white and black site lists and compare them
+            # to one we received from upstream service (ReqMgr2)
+            wmaWhiteList = wHelper.getSiteWhiteList()
+            wmaBlackList = wHelper.getSiteBlackList()
+            if wmaWhiteList != siteWhiteList or wmaBlackList != siteBlackList:
+                self.logger.info(f"Updating {wflow}: siteWhiteList {wmaWhiteList} => {siteWhiteList} and siteBlackList {wmaBlackList} => {siteBlackList}")
+                try:
+                    # update local WorkQueue first
+                    self.localWQ.updateSiteLists(wflow, siteWhiteList, siteBlackList)
+                except Exception as ex:
+                    msg = f"Caught unexpected exception in SiteListUpdater. Details:\n{str(ex)}"
+                    logging.exception(msg)
+                    raise LocalWorkqueueUpdateException(msg) from None
```

Comment:
If we raise this exception, we end up breaking the cycle. Instead, we should record the error but proceed with the other workflows in the pipeline.

Reply:
ok, switched to continue